### ISSUE TITLE
What is the proper way to deserialize a List of things?

### ISSUE THREAD

**Issue by deiga** [kittinunf/Fuel#233]
I've a data class and a request which returns a list of these objects. What is the proper way to deserialize the list? I don't want to write an explicit deserializer just for `List<CommitStatus>`

I tried to do automatic deserialization with `.httpGet().responseObject<List<CommitStatus>> { ... }` It returns a list of things nicely, but then it fails when accessing the list elements. 
```
java.lang.ClassCastException: com.google.gson.internal.LinkedTreeMap cannot be cast to io.smartly.balex.receiver.CommitStatus
``` 

This is my data class:
```
data class CommitStatus(val state: String, val context: String, val target_url: String) {
    class Deserializer : ResponseDeserializable<CommitStatus> {
        override fun deserialize(content: String): CommitStatus? {
            return Gson().fromJson(content, CommitStatus::class.java)
        }
    }
}
```

**Comment by yoavst** [comment]
Your deserializer is not actually used. If you use `fuel-gson`, you use the built-in deserializer.

Can you try it with `responseObject<ArrayList<CommitStatus>>`? I had problems even in Java that deserializing directly to `List` fails but to an implementation works.

**Comment by deiga** [comment]
@yoavst Yeah, I know that. I tried using `responseObject(CommitStatus.Deserializer(), ...) ` But that didn't yield a list.

Nope, I'm quite sure the response is an JSON array with objects

**Comment by yoavst** [comment]
Could you paste here a json response so I could try to reproduce it?

**Comment by deiga** [comment]
The `ArrayList<..>` didn't work

Here's an example response JSON https://gist.github.com/deiga/50ddc4bb4384946ca5029d8b7b65f0ad

**Comment by yoavst** [comment]
Seems like the problem is type erasure.  Working on a fix.

**Comment by kittinunf** [comment]
@deiga Thanks for reporting this.

Hmm I think I use `Fuel` to work with List of items before. Can we do something like this?

```
data class Issue(
            val id: Int = 0,
            val number: Int = 0,
            val title: String = "",
    ) {
        class Deserializer : ResponseDeserializable<Issue> {
            override fun deserialize(reader: Reader) = Gson().fromJson(reader, Issue::class.java)
        }

        class ListDeserializer : ResponseDeserializable<List<Issue>> {
            override fun deserialize(reader: Reader): List<Issue> {
                val type = object : TypeToken<List<Issue>>() {}.type
                return Gson().fromJson(reader, type)
            }
        }
    }
```

Usage

```
"https://api.github.com/repos/kittinunf/Fuel/issues".httpGet()
                .responseObject(Photo.ListDeserializer()) { _, _, result ->
                    update(result)
                }
```

**Comment by kittinunf** [comment]
I have also added PR to add this into sample module, please kindly take a look here

https://github.com/kittinunf/Fuel/pull/236

**Comment by yoavst** [comment]
@kittinunf #235 will fix it with no change of code from users.

**Comment by kittinunf** [comment]
I understand. But this is not a bug per se, probably a feature request. 

Also, in #235, it is probably only valid for Gson module, not the for the core Fuel module right?

**Comment by yoavst** [comment]
It's actually a bug since it's unexpected behavior and causes crashes.

If you provide generic class deserializer to core fuel it may also cause crash.

**Comment by deiga** [comment]
@kittinunf The problem I have with #235 is that building a special `ListDeserializer` seems a bit silly, since any Deserialization library should be able to deserialize Collections automatically

**Comment by SleeplessByte** [comment]
This is fixed properly with 2.0.0
